Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp

```diff
@@ -3850,6 +3854,225 @@ auto AirIRGenerator::addCallRef(const TypeDefinition& signature, Vector<Expressi
+    const auto& callingConvention = wasmCallingConvention();
+    CallInformation wasmCallerInfo = callingConvention.callInformationFor(callerTypeDefinition, CallRole::Callee);
+    CallInformation wasmCalleeInfo = callingConvention.callInformationFor(signature, CallRole::Caller);
+
+    m_makesCalls = true;
+
+    for (unsigned i = 0; i < signature.as<FunctionSignature>()->returnCount(); ++i)
+        results.append(tmpForType(signature.as<FunctionSignature>()->returnType(i)));
+
+    Vector<UnlinkedWasmToWasmCall>* unlinkedWasmToWasmCalls = &m_unlinkedWasmToWasmCalls;
+
+    if (m_info.isImportedFunctionFromFunctionIndexSpace(functionIndex)) {
+        m_maxNumJSCallArguments = std::max(m_maxNumJSCallArguments, static_cast<uint32_t>(args.size()));
+
+        auto currentInstance = g64();
+        append(Move, instanceValue(), currentInstance);
+
+        auto targetInstance = g64();
+
+        // FIXME: We should have better isel here.
+        // https://bugs.webkit.org/show_bug.cgi?id=193999
+        append(Move, Arg::bigImm(Instance::offsetOfTargetInstance(functionIndex)), targetInstance);
+        append(Add64, instanceValue(), targetInstance);
+        append(Move, Arg::addr(targetInstance), targetInstance);
+
+        BasicBlock* isWasmBlock = m_code.addBlock();
+        BasicBlock* isEmbedderBlock = m_code.addBlock();
+        BasicBlock* continuation = m_code.addBlock();
+
+        append(BranchTest64, Arg::resCond(MacroAssembler::NonZero), targetInstance, targetInstance);
+        m_currentBlock->setSuccessors(isWasmBlock, isEmbedderBlock);
+
+        {
+            auto pair = emitCallPatchpoint(isWasmBlock, signature, results, args);
+            auto* patchpoint = pair.first;
+            auto exceptionHandle = pair.second;
+            // We need to clobber all potential pinned registers since we might be leaving the instance.
+            // We pessimistically assume we could be calling to something that is bounds checking.
+            // FIXME: We shouldn't have to do this: https://bugs.webkit.org/show_bug.cgi?id=172181
+            patchpoint->clobberLate(PinnedRegisterInfo::get().toSave(MemoryMode::BoundsChecking));
+
+            patchpoint->setGenerator([=, this](CCallHelpers& jit, const B3::StackmapGenerationParams& params) {
+                AllowMacroScratchRegisterUsage allowScratch(jit);
+                prepareForReturnCall(jit, wasmCallerInfo, wasmCalleeInfo);
+                exceptionHandle.generate(jit, params, this);
+                CCallHelpers::Call call = jit.threadSafePatchableNearTailCall();
+                jit.addLinkTask([unlinkedWasmToWasmCalls, call, functionIndex](LinkBuffer& linkBuffer) {
+                    unlinkedWasmToWasmCalls->append({ linkBuffer.locationOfNearCall<WasmEntryPtrTag>(call), functionIndex });
+                });
+            });
+
+            append(isWasmBlock, Jump);
+            isWasmBlock->setSuccessors(continuation);
```

Comment:
We shouldn't be telling the compiler about control flow that doesn't exit. Let's make the Patchpoint terminal, and then we don't need this Jump control flow edge.